### ISSUE TITLE
Translation error message for variables used in DynamicSelect statements also for partial classes

### ISSUE THREAD

**Issue by dietmarw** [OpenModelica/OpenModelica#13092]
### Description

The OpenHPL libray makes use of [DynamicSelect for the diagram graphics](https://github.com/OpenSimHub/OpenHPL/blob/master/OpenHPL/Icons/Surge.mo#L34-L35). Event though this is not yet fully supported in OMEdit it should not trigger a Translation Error message.

### Steps to Reproduce

1. open `OpenHPL.Examples.Simple`
2. When the diagram is rendered in OMEdit the following error is reported:
```log
[1] 10:36:25 Translation Error
[OpenHPL.Icons.Surge: 5:57-47:79]: Variable l not found in scope Surge.
```

### Expected Behavior

No error message should be reported though the variable `l` is not present in `OpenHPL.Icons.Surge`, that class is a partial class and does not need to have the variable. 

### Version and OS

 - OpenModelica Version: 1.25.0~dev-114
 - OS: Linux
 - Versions of used Modelica libraries if applicable: OpenHPL 2.0.1

**Comment by perost** [comment]
The specification does say that a partial class is allowed to be incomplete, though without going into any details about what that means. But I'm confused about what the intention is here, since there's no way a user can add the missing variables to `Surge` except by making their own copy of the class and editing it. Surely that's not the intended use of the class?

**Comment by perost** [comment]
> The intended use here is that it is an icon class that is used by the model that then provides the necessary parameter and variable.

How? A model can't add variables to a class it extends.

**Comment by perost** [comment]
> Sure it can. That's why we have `partial`

`partial` just means that the ways the class can be used is restricted, such as not being possible to instantiate. It doesn't change how e.g. `extends` works.

**Comment by dietmarw** [comment]
Partial also means that you do not have to have a balanced model, i.e., variables can be missing which then are provided by the classes extending from the partial base class.

**Comment by perost** [comment]
> Partial also means that you do not have to have a balanced model, i.e., variables can be missing which then are provided by the classes extending from the partial base class.

I think you might be confused about how extends works. I.e. this is not legal:
```Modelica
partial model A
equation
  x = 1; // Error, no x in this scope.
end A;

model B
  extends A;
  Real x;
end B;
```
But this is:
```Modelica
// A is not balanced.
partial model A
  Real x;
end A;

// B is balanced.
model B
  extends A;
equation
  x = 1;
end B;
```

**Comment by dietmarw** [comment]
Right, I see. OK in that case the question is, do these rules also apply to annotations and specifically to DynamicSelect. Dymola does not think so, hence I never got an error message there regarding this. They also have full support for the implementation of DynamicSelect.

**Comment by perost** [comment]
> Right, I see. OK in that case the question is, do these rules also apply to annotations and specifically to DynamicSelect. Dymola does not think so, hence I never got an error message there regarding this. They also have full support for the implementation of DynamicSelect.

You might not have gotten an error message, but does it actually work in Dymola or does Dymola just ignore the error? As far as I know there aren't any special lookup rules for annotations or `DynamicSelect` that would make something like this legal.

**Comment by dietmarw** [comment]
It does work in Dymola :-)
![surgetank](https://github.com/user-attachments/assets/fba53d78-b40c-47be-abdc-de89c255fa5e)

**Comment by perost** [comment]
> It does work in Dymola :-)

Ok, I can't find anything in the specification that says that it should work. So it seems like it's either a bug in Dymola or something that needs to be added to the specification.

I don't think I've ever seen any other library do it. Icon classes that use `DynamicSelect` usually just do it the normal way of declaring variables in the icon class that the extending class then add binding equations to via modifiers.

**Comment by casella** [comment]
@dietmarw we have something like this in [PowerGrids 2.0.0](https://github.com/PowerGrids/PowerGrids/tree/develop), see e.g. the [OnePortDynamicText](https://github.com/PowerGrids/PowerGrids/blob/8d68fd7d0bd995c9c7dacbc27f76e26928c8c606/PowerGrids/Icons.mo#L91) model. 

This is not even a partial model: it has input variables, which are then set by modifiers when adding the extends clause to the model that wants to use the icon with DynamicSelect stuff. You may follow the same pattern.

**Comment by dietmarw** [comment]
@casella Thanks. I know how to work around this, the only question is if it is actually necessary.

**Comment by casella** [comment]
I also always understood `partial` as Per does, i.e. "incomplete" means missing equations, but not missing variables declarations. It seemed obvious to me that if you used a variable in a model (even in an annotation), it must have been declared.

On the other hand, it is also true, as @perost notices, that the MLS is a bit vague [there](https://specification.modelica.org/maint/3.6/class-predefined-types-and-declarations.html#component-declaration-static-semantics):
> A class defined with **partial** in the _class-prefixes_ is called a _partial_ class. Such a class is allowed to be incomplete, and cannot be instantiated in a simulation model; useful, e.g., as a base class.
as it is not really clear what "incomplete means". Maybe we should ask for clarification in the MAP-Lang issue tracker.

I made a simple MWE:

```modelica
package TestPartial
  partial model A
    Real x = p;
  end A;

  model B
    extends A;
    parameter Real p = 10;
  end B;
  
  partial model C
  annotation(
      Icon(graphics = {Ellipse(origin = {0, 0}, extent = {{-p, p}, {p, -p}})}));
  end C;
  
  model D
    extends C;
    parameter Real p = 10;
  end D;
  
  partial model E
  annotation(
      Icon(graphics = {Ellipse(origin = {0, 0}, 
                               extent = DynamicSelect({{- 10, 10}, {10, -10}},{{-r, r}, {r, -r}}))}));  
  end E;

  model F
    extends E;
    Real r = 10*(1 + 5*time);
  end F;
end TestPartial;
```
Models A and B, which use undeclared stuff in binding equations, are rejected both by Dymola and OMC.

Models C, D, E, and F are checked successfully both by Dymola and OMC

Models D and E, which are not partial, can be successfully run both in Dymola and OMC.

However, when I open the package in OMEdit and expand it, I get
```modelica
[3] 13:39:50 Scripting Error
[D:/Temp/OMEdit/TestPartial.mo:13:12-13:78:writable] Error: Variable p not found in scope C.

[4] 13:39:50 Scripting Error
[D:/Temp/OMEdit/TestPartial.mo:13:12-13:78:writable] Error: Variable p not found in scope C.

[5] 13:39:50 Translation Error
[TestPartial: 23:12-24:100]: [Variable r not found in scope E.](omeditmessagesbrowser:///TestPartial?lineNumber=23)

[6] 13:39:50 Translation Error
[TestPartial: 23:12-24:100]: [Variable r not found in scope E.](omeditmessagesbrowser:///TestPartial?lineNumber=23)
```
because apparently the inst-based API is not happy with models C and E.

I think the main question is which of those models should be legal. Then we can implement that. I'll open a ticket on the MLS tracker later today on the subject

**Comment by perost** [comment]
> However, when I open the package in OMEdit and expand it, I get
> 
> ```modelica
> [3] 13:39:50 Scripting Error
> [D:/Temp/OMEdit/TestPartial.mo:13:12-13:78:writable] Error: Variable p not found in scope C.
> 
> [4] 13:39:50 Scripting Error
> [D:/Temp/OMEdit/TestPartial.mo:13:12-13:78:writable] Error: Variable p not found in scope C.
> 
> [5] 13:39:50 Translation Error
> [TestPartial: 23:12-24:100]: [Variable r not found in scope E.](omeditmessagesbrowser:///TestPartial?lineNumber=23)
> 
> [6] 13:39:50 Translation Error
> [TestPartial: 23:12-24:100]: [Variable r not found in scope E.](omeditmessagesbrowser:///TestPartial?lineNumber=23)
> ```
> 
> because apparently the inst-based API is not happy with models C and E.

The reason for this is that we don't care about graphical annotations when checking or compiling models, so the compiler just ignores an annotations not directly related to the simulation. But when you open the package in OMEdit we obviously want to use them to display e.g. icons, in which case the missing variables become a bit problematic.

**Comment by casella** [comment]
See discussion in modelica/ModelicaSpecification#3598. You may add your comments there.

**Comment by casella** [comment]
@dietmarw I checked the discussion in modelica/ModelicaSpecification#3598, which is now closed. As I understand, the consensus is that lookup rules should always apply. So, nothing in a partial model can refer to undeclared variables for the simple reason that looup would fail.

I would consider this issue resolved. Please reopen and motivate why if you think otherwise.

**Comment by dietmarw** [comment]
Fair enough.
